SCAutolib/controller.py

```diff
@@ -221,136 +275,140 @@ def setup_user(self, user_dict: dict, force: bool = False):
+            if new_card.user.user_type == UserType.local:
+                new_card.cnf = self.prepare_user_cnf(new_card)
+            if force:
+                self.revoke_certs(new_card)
+            new_card.create()
+        else:
+            raise NotImplementedError("Other card types than 'physical' and "
+                                      "'virtual' are not supported")
+
+        dump_to_json(new_card)
+        return new_card
+
+    def link_user_to_card(self, card: card.VirtualCard):
+        for card_user in self.users:
+            if card_user.username == card.cardholder:
+                return card_user
+
+    def prepare_hsm(self, card_dir: Path = None):
+        """Prepare HSM config for virtual card"""
+        filepath = card_dir.joinpath("sofhtsm2.conf")
+        hsm_conf = file.SoftHSM2Conf(filepath, card_dir=card_dir)
+        hsm_conf.create()
+        hsm_conf.save()
+        return hsm_conf
+
+    def prepare_user_cnf(self, card: card.VirtualCard):
+        """Prepare user openssl cnf"""
+        cnf_path = card.card_dir.joinpath(f"{card.cardholder}.cnf")
+        cnf = file.OpensslCnf(filepath=cnf_path, conf_type="user",
+                              replace=[card.cardholder, card.CN])
+        cnf.create()
+        cnf.save()
+        return cnf.path
+
+    def revoke_certs(self, card: card.VirtualCard):
+        if card.cert and card.cert.exists():
+            if card.user.user_type == UserType.local:
+                self.local_ca.revoke_cert(card.cert)
+            else:
+                self.ipa_ca.revoke_cert(card.cert)
+
+    def enroll_card(self, card: card.VirtualCard):
+        """
+        Enroll the card - i.e. upload keys and certs to card. If private key
+        and/or the certificate do not exist, new one's would be requested
+        from corresponding CA.
 
-            csr = user_.gen_csr()
+        :param card: card object
+        :type card: card.VirtualCard
+        """
+        logger.debug(f"Starting enrollment of the card {card.name}")
+        if not card:
+            raise exceptions.SCAutolibException(
+                f"Card {card.name} is not initialized")
 
-            ca = self.ipa_ca \
-                if isinstance(user_, user.IPAUser) else self.local_ca
-            user_.cert = ca.request_cert(csr, user_.username, user_.cert)
+        if not card.key.exists():
+            _gen_private_key(card.key)
 
-        user_.card.enroll()
+        if not card.cert.exists():
+            csr = card.gen_csr()
+            ca = self.ipa_ca \
+                if isinstance(card.user, user.IPAUser) else self.local_ca
+            card.cert = ca.request_cert(csr, card.cardholder, card.cert)
 
-        dump_to_json(user_.card)
-        # Update the dump for current user to link it with the card
-        dump_to_json(user_)
+        card.enroll()
+        dump_to_json(card)
 
-    def cleanup(self):
+    def cleanup(self):  # FIXME
```

Comment:
its a good practice to describe the fixme with at least one sentence so the next reader understand what is the issue and what needs to be fixed (even if not how).